src/v/cloud_storage/remote_segment_index.cc

```diff
@@ -28,11 +31,18 @@ offset_index::offset_index(
   , _initial_file_pos(initial_file_pos)
   , _rp_index(initial_rp)
   , _kaf_index(initial_kaf)
-  , _file_index(initial_file_pos) {}
+  , _file_index(initial_file_pos)
+  , _file_pos_sampling_step(sampling_step) {}
 
 void offset_index::add(
   model::offset rp_offset, model::offset kaf_offset, int64_t file_offset) {
     auto ix = index_mask & _pos++;
+    vassert(
+      file_offset >= _file_pos_sampling_step,
+      "file offset {} is smaller than the sampling step {}",
+      file_offset,
+      _file_pos_sampling_step);
+    file_offset -= _file_pos_sampling_step;
```

Comment:
Correct me if I'm wrong, but I think we are supposed to subtract the sampling step from deltas, not from the absolute offset. In the latter case we simply create a shift, but deltas are still large and will have to be encoded using 2 bytes.

Reply:
fixed that, also reworked the PR quite a bit

Reply:
> Correct me if I'm wrong, but I think we are supposed to subtract the sampling step from deltas, not from the absolute offset. In the latter case we simply create a shift, but deltas are still large and will have to be encoded using 2 bytes.

Should this have been caught in a test and if so did that test get added?

Reply:
I guess we can devise a lower bound on compression ratio based on the range of random encoded values and check that.